CI,TYP: Bump mypy to 1.4.1

BvB93 opened the pull request:
This PR bumps the mypy version used throughout the test suite from 0.981 to 1.3.0 (it's been a while...).

In addition to mypy now having implemented support for `typing.Concatenate` there seems been some changes in how it deals with the joining of super- and sub-classes. The end result as that, _e.g._ the likes of `np.float64` + `np.float32` were previously interpretted as producing `np.floating[_64Bit]` while not it results in a `np.floating[_64Bit | _32Bit]` union. 

While non-ideal, this upstream change is admitetly a minor inconvenience in the grand scheme of things, mostly because the typing of dtype bitsizes in numpy is severely lacking due to the sheer number of challenges associated with it.

Lastly, while not implemented in this PR there's also [`typing.Self`](https://docs.python.org/3/library/typing.html#typing.Self) support ever since mypy 1.0.0. This type is essentially a bit of syntactic sugar to make typing involving `self` and `cls` a bit easier, removing their need for hyper-specialized typevars. Something for a future PR perhaps.

BvB93 committed:
CI,TYP: Bump mypy to 1.3.0

BvB93 committed:
TYP: Improve function-based annotations with `typing.Concatenate`

BvB93 commented:
```
>               pytest.fail(f"Unexpected mypy standard error\n\n{stderr}")
E               Failed: Unexpected mypy standard error
E               
E               mypy: "../../venv-for-wheel/lib/python3.9/site-packages/mypy_extensions.py" shadows library module "mypy_extensions"
E               note: A user-defined top-level module with name "mypy_extensions" is not supported
```
Some more debugging for tomorrow. The error messages differ a bit from platform to platform, though this seems to suggest that a mypy from outside `venv-for-wheel/lib/python3.9/...` is clashing with a `mypy_extensions` installation inside the venv environment. Thus far I've been unable to reproduce the issue locally though (be it either in mac or windows) with a simple `python .\runtests.py -t numpy/typing --mode=full`.

Any thoughts?

seberg commented:
A quick google search finds this: https://stackoverflow.com/questions/64806361/installing-mypy-leads-to-shadows-library-module-error

The travis-CI run sets the `MYPYPATH` to the python path, which is strongly discouraged (and I guess mypy broke it entirely?).  Sounds like fixing `MYPYPATH` to just include numpy should do the trick?

BvB93 committed:
CI: Avoid overriding `MYPYPATH`

This practice is actively discouraged by mypy

bananarama92 commented:
It did solve the Travis failure though; and the fact that all other runs that had previously failed do so again do make me suspect that there's more going on here than a simple caching problem.

BvB93 commented:
It did solve the Travis failure though; and the fact that all other runs that had previously failed do so again do make me suspect that there's more going on here than a simple caching problem.

seberg commented on test_requirements.txt:
```diff
@@ -10,7 +10,7 @@ cffi; python_version < '3.10'
 # NOTE: Keep mypy in sync with environment.yml
-mypy==0.981; platform_python_implementation != "PyPy"
+mypy==1.3.0; platform_python_implementation != "PyPy"
 typing_extensions>=4.2.0
```
```suggestion
typing_extensions>=4.2.0; python_version < '3.10'
```
maybe?

rgommers replied:
this comment still looks valid, because `typing_extensions` is used for Python 3.9 still. and it's also in `test_requirements.txt`

BvB93 commented:
Hm, not sure about `typing_extensions` lacking. Pretty much all the errors mention either a C/D path mismatch or, more directly, that either the mypy_extensions or typing_extension in the testing venv is shadowing wherever the package was installed originally.

BvB93 commented:
Think I may have found something here; seems the problem might be that it is directly run against something in site-packages: https://github.com/python/mypy/issues/11477. It's annoying that this has now become an issue all of a sudden, though I wonder if we could run mypy directly against the numpy source? The .pyi files aren't affected by the build stage after all.

BvB93 commented:
Marking as draft for now until I have time to take a closer look.

rgommers commented:
> It's annoying that this has now become an issue all of a sudden, though I wonder if we could run mypy directly against the numpy source? The .pyi files aren't affected by the build stage after all.

It should sorta work I think; the exceptions like generated `.py`/`.pyi` files are corner cases where there may be issues. The tests under `numpy/typing/` import numpy though, so they'll work with editable installs and `spin test`, but not with a regular installed package.

That Mypy issue is a bit sad of course; they can/should just fix the thing where `typing_extensions` being installed breaks things for them.

rgommers commented:
I tried to add a Mypy run with Python 3.12.0rc1 (as part of gh-24291), but it chokes on:
```
 ERROR numpy/typing/tests/test_typing.py - DeprecationWarning: ast.Bytes is deprecated and will be removed in Python 3.14; use ast.Constant instead
```
I suspect that this Mypy upgrade will fix that, so I haven't bothered to look at it in more detail.

BvB93 committed:
TYP: Bump mypy to 1.5.1

BvB93 committed:
TYP: Lower the mypy pin to 1.4.1

BvB93 committed:
TYP,TST: Adapt the typing test suite to mypy >=1.4 changes

BvB93 commented:
Right, have yet to get mypy 1.5.1 to cooperate, but 1.4.1 now works making this PR good to go. 

Mypy 1.5.1 seems to truncate its type evaluation before all `fail`-related tests are completed, which causes issue. I suspect that this is related to the new "Consistently Avoid Type Checking Unreachable Code" improvements ([1.5.1 changelog](https://mypy-lang.blogspot.com/2023/08/mypy-15-released.html)); hopefully this is something that can be fixed by implementing https://github.com/numpy/numpy/issues/21505 via useage of `typing.assert_type` and `.assert_never`.

rgommers reviewed (approved):
This looks great, thanks @BvB93. After going through the changes, the one thing that stood out to me was the "Union of scalars" thing which looks like a regression. But then I saw that you already explained it:

> In addition to mypy now having implemented support for `typing.Concatenate` there seems been some changes in how it deals with the joining of super- and sub-classes. The end result as that, _e.g._ the likes of `np.float64` + `np.float32` were previously interpreted as producing `np.floating[_64Bit]` while not it results in a `np.floating[_64Bit | _32Bit]` union.
>
> While non-ideal, this upstream change is admitetly a minor inconvenience in the grand scheme of things, mostly because the typing of dtype bitsizes in numpy is severely lacking due to the sheer number of challenges associated with it.

Fair enough, I agree it's a minor thing and we can simply ignore it for now.

I'll put back the `typing_extensions` requirement in `environment.yml` and then this should be good to go.

rgommers committed:
TYP: guard `typing_extensions` usage for <py310, add it back in environment.yml

It's still needed, so keep it in the environment. The added import
guards here match the ones already present in two other `.pyi` files.

[skip cirrus] [skip circle] [skip azp] [skip travis]

rgommers commented:
I found another small issue with the use of `typing_extensions` - pushed a fix.

rgommers commented:
Okay, in it goes. Thanks Bas!
